File: DailyChallenges/Scala/LeetCode_2187.scala
object Solution {
    def minimumTime(time: Array[Int], totalTrips: Int): Long = {
        def count(t: Long): Boolean = {
            time.foldLeft(0L)((c, i) => {
                val _c = c + t/i;
                if (_c >= totalTrips) return false;
                _c
            })
            true
        }

        def bs(l: Long, r: Long): Long = if (l >= r) l else {
            val m = l + (r-l)/2
            if (count(m)) bs(m+1, r) else bs(l, m)
        }

        bs(1, 1e14.toLong)
    }
}

File: DailyChallenges/CPP/LeetCode_2187.cpp
class Solution {
public:
    long long minimumTime(vector<int>& time, int totalTrips) {
        function<bool(long long)> count = [&](long long t) {
            long long ans = 0;
            for (int pt: time){
                ans += t/pt;
                if (ans >= totalTrips) return false;
            }
            return true;
        };

        long long l = 1LL, r = 1e14;
        while (l < r){
            long long m = l+(r-l)/2;
            if (count(m)) l = m+1; else r = m; 
        }

        return l;
    }
};
